src/ORM/DataList.php

```diff
@@ -930,50 +932,9 @@ private function setDataObjectEagerLoadedData(DataObject $item): void
     {
         // cache $item->ID at the top of this method to reduce calls to ViewableData::__get()
         $itemID = $item->ID;
-        foreach (array_keys($this->eagerLoadedData) as $eagerLoadRelation) {
-            list($dataClasses, $relations) = $this->getEagerLoadVariables($eagerLoadRelation);
-            $dataClass = $dataClasses[count($dataClasses) - 2];
-            $relation = $relations[count($relations) - 1];
-            foreach (array_keys($this->eagerLoadedData[$eagerLoadRelation]) as $eagerLoadID) {
-                $eagerLoadedData = $this->eagerLoadedData[$eagerLoadRelation][$eagerLoadID][$relation];
-                if ($dataClass === $dataClasses[0]) {
-                    if ($eagerLoadID === $itemID) {
-                        $item->setEagerLoadedData($relation, $eagerLoadedData);
-                    }
-                } elseif ($dataClass === $dataClasses[1]) {
-                    $relationData = $item->{$relations[1]}();
-                    if ($relationData instanceof DataObject) {
-                        if ($relationData->ID === $eagerLoadID) {
-                            $subItem = $relationData;
-                        } else {
-                            $subItem = null;
-                        }
-                    } else {
-                        $subItem = $item->{$relations[1]}()->find('ID', $eagerLoadID);
-                    }
-                    if ($subItem) {
-                        $subItem->setEagerLoadedData($relations[2], $eagerLoadedData);
-                    }
-                } elseif ($dataClass === $dataClasses[2]) {
-                    $relationData = $item->{$relations[1]}();
-                    if ($relationData instanceof DataObject) {
-                        $list = new ArrayList([$relationData]);
-                    } else {
-                        $list = $relationData;
-                    }
-                    foreach ($list as $subItem) {
-                        $subRelationData = $subItem->{$relations[2]}();
-                        if ($relationData instanceof DataObject) {
-                            $subList = new ArrayList([$subRelationData]);
-                        } else {
-                            $subList = $subRelationData;
-                        }
-                        $subSubItem = $subList->find('ID', $eagerLoadID);
-                        if ($subSubItem) {
-                            $subSubItem->setEagerLoadedData($relations[3], $eagerLoadedData);
-                        }
-                    }
-                }
+        foreach (array_keys($this->eagerLoadedData) as $relation) {
+            if (array_key_exists($itemID, $this->eagerLoadedData[$relation])) {
+                $item->setEagerLoadedData($relation, $this->eagerLoadedData[$relation][$itemID][$relation]);
```

Comment:
I've restructured the `eagerLoadedData` array so that it only contains the _top_ level of relational data. Sub relations are now being added directly to the appropriate parent `EagerLoadedDataList` or `DataObject` record as part of the `fetchEagerLoadRelations()` implementation.